### ISSUE TITLE
Getting the training set from a certain crossvalidation iteration

### ISSUE THREAD

**Issue by HeidiSeibold** [mlr-org/mlr#1490]
This is a question that I have because I am using OpenML+mlr to test the new `ctree` function. 

```
library("OpenML")
source("https://raw.githubusercontent.com/HeidiSeibold/sandbox/master/rstuff/openml_newctree/new_ctree_mlr.R")

tsk <- getOMLTask(task.id = 2285)
rn <- runTaskMlr(task = tsk,
                 learner = makeLearner("regr.develpartykit.ctree", multiway = TRUE),
                 seed = 1)
```

gives me an Error 

```
> rn <- runTaskMlr(task = tsk,
+                  learner = makeLearner("regr.develpartykit.ctree", multiway = TRUE),
+                  seed = 1)
Task: cleveland, Learner: regr.develpartykit.ctree
[Resample] cross-validation iter: 1
[Resample] cross-validation iter: 2
[Resample] cross-validation iter: 3
 Error in kids_node(node)[[i]] : subscript out of bounds 
```

Now I want to extract cross validation iteration 3, since the error happens there. I know that the problem occurs while predicting. I thought the following code should result in the same error:

```
data <- tsk$input$data.set$data
cv <- tsk$input$estimation.procedure$data.splits
fold <- 3
rows_train <- cv$rowid[cv$fold == fold & cv$type == "TRAIN"]
rows_test <- cv$rowid[cv$fold == fold & cv$type == "TEST"]
d_train <- data[rows_train, ]
d_test <- data[rows_test, ]
library("partykit")
set.seed(1)
ctr <- ctree(num ~ ., data = d_train, 
             control = ctree_control(multiway = TRUE))
predict(ctr, newdata = d_test)
```
But I don't get an error. Any ideas what I am doing wrong? During debugging I found that the test set is the set where the error happens but the tree is different (apparently mlr uses a different training set than I do).

**Comment by berndbischl** [comment]
1) please always try to post a traceback

2) i can try to take a brief look now

**Comment by HeidiSeibold** [comment]
> 1. please always try to post a traceback
It was very long, that's why I didn't

> 2. i can try to take a brief look now
Thanks :)
I guess the question is just, whether
```
rows_train <- cv$rowid[cv$fold == fold & cv$type == "TRAIN"]
d_train <- data[rows_train, ]
```
returns the training set of a certain cross validation fold.

**Comment by HeidiSeibold** [comment]
I think I found what makes the difference. I do not yet understand why

- This one throws the error:
```
set.seed(1)
lrnr <- trainLearner.classif.develpartykit.ctree(.learner = makeLearner("regr.develpartykit.ctree", multiway = TRUE),
                                                 .task = convertOMLTaskToMlr(tsk)$mlr.task, 
                                                 .subset = rows_train,
                                                 multiway = TRUE)
predict(lrnr, newdata = d_test)
```

- This one does not:
```
set.seed(1)
lrnr <- trainLearner.classif.develpartykit.ctree(.learner = makeLearner("regr.develpartykit.ctree", multiway = TRUE),
                                                 .task = convertOMLTaskToMlr(tsk)$mlr.task, 
                                                 .subset = rows_train)
predict(lrnr, newdata = d_test)
```

The only difference between the two is the additional `multiway = TRUE` in the first one.

**Comment by HeidiSeibold** [comment]
Now this is surprising:

- `set.seed` after `getTaskData`:

```
> dd <- getTaskData(task = convertOMLTaskToMlr(tsk)$mlr.task, 
+                   subset = rows_train)
> set.seed(1)
> ctr <- ctree(num ~ ., data = dd, 
+              control = ctree_control(multiway = TRUE))
> predict(ctr, newdata = d_test)
       156        265        270         73        225        122         21        150         31         22        201        221 
0.76315789 0.00000000 2.17460317 1.21621622 0.01886792 0.29411765 0.01886792 0.76315789 2.17460317 0.29411765 0.29411765 0.01886792 
        64         15         88         36         54         81         96         30         40        261        169        116 
2.17460317 0.01886792 0.29411765 2.55555556 2.17460317 0.29411765 2.17460317 2.17460317 2.17460317 0.06666667 0.29411765 0.01886792 
       206        155        152        117        211         78        292 
2.17460317 2.17460317 0.76315789 0.01886792 0.76315789 0.01886792 0.00000000 
```


- `set.seed` before `getTaskData`:

```
> set.seed(1)
> dd <- getTaskData(task = convertOMLTaskToMlr(tsk)$mlr.task, 
+                   subset = rows_train)
> ctr <- ctree(num ~ ., data = dd, 
+              control = ctree_control(multiway = TRUE))
> predict(ctr, newdata = d_test)
Error in kids_node(node)[[i]] : subscript out of bounds
```

**Why does `getTaskData` have an impact on the seed?**

**Comment by jakob-r** [comment]
I would guess it is `convertOMLTaskToMlr`. Try to call it independently beforehand. You can check the values of `.Random.seed` to see if something changed.

**Comment by HeidiSeibold** [comment]
You are right

```
> set.seed(1)
> str(.Random.seed)
 int [1:626] 403 624 -169270483 -442010614 -603558397 -222347416 1489374793 865871222 1734802815 98005428 ...
> tsk_mlr <- convertOMLTaskToMlr(tsk)$mlr.task
> str(.Random.seed)
 int [1:626] 403 607 1654269195 -1877109783 -961256264 1403523942 124639233 261424787 1836448066 1034917620 ...
> dd <- getTaskData(task = tsk_mlr, 
+                   subset = rows_train)
> str(.Random.seed)
 int [1:626] 403 607 1654269195 -1877109783 -961256264 1403523942 124639233 261424787 1836448066 1034917620 ...
```
I guess I'll write an OpenML issue instead...

**Comment by HeidiSeibold** [comment]
Thanks for the help @jakob-r and for the `.Random.seed` tip. I had tried to google how to print the seed but hadn't been very successful.

**Comment by giuseppec** [comment]
The problem is not OpenML is the following (therefore I'll reopen here):

- `convertOMLTaskToMlr` calls internally `removeConstantFeatures`.
- `mlr:::removeConstantFeatures.data.frame` calls again `BBmisc::computeMode` with it's default `ties.method = "random"` that changes the random seed:

```
> set.seed(1)
> str(.Random.seed)
 int [1:626] 403 624 -169270483 -442010614 -603558397 -222347416 1489374793 865871222 1734802815 98005428 ...
> m = computeMode(1:4) # by default it uses ties.method = "random"
> str(.Random.seed)
 int [1:626] 403 1 1654269195 -1877109783 -961256264 1403523942 124639233 261424787 1836448066 1034917620 ...
```

If we don't want this behaviour, we should either directly change the default in `BBmisc::computeMode` or use `computeMode(x, ties.method = "first")` in mlr's `mlr:::removeConstantFeatures.data.frame`.

**Comment by berndbischl** [comment]
> Now I want to extract cross validation iteration 3, since the error happens there.

do you still have that question? because the simplest approach is:
convert the task to mlr. 
that gives you a ResampleInstance
that contains training and test sets

another possibly simpler option is to use 
on.learner.error = warn
then you can completely execute the resampling
and look at all results post-hoc

@HeidiSeibold are there any problems still open? can you give me feedback

@HeidiSeibold @giuseppec 
if you want removeConstantFeatures changed please open a clean new issue for that

**Comment by HeidiSeibold** [comment]
> @HeidiSeibold are there any problems still open? can you give me feedback

No I think the help I got was great. 

> @HeidiSeibold @giuseppec
if you want removeConstantFeatures changed please open a clean new issue for that

The problems I had were due to the fact that it was kind of unexpected for me that `convertOMLTaskToMlr` changes the seed. There are in my opinion differnt scenarios how to solve this

- Change the default of `removeConstantFeatures` to something not impacting the seed.
- Document in `runTaskMlr` that the seed the user sets is not for the algorithm but also for the data preparation and hint to the randomness in `removeConstantFeatures`.
- Decide that this usecase of mine is not happening very often and can be ignored.

**Comment by berndbischl** [comment]
> Change the default of removeConstantFeatures to something not impacting the seed.

i think we can do that.
